src/librustc_mir/dataflow/impls/borrows.rs

```diff
@@ -46,9 +47,118 @@ pub struct Borrows<'a, 'gcx: 'tcx, 'tcx: 'a> {
     root_scope: Option<region::Scope>,
 
     borrow_set: Rc<BorrowSet<'tcx>>,
+    borrows_out_of_scope_at_location: FxHashMap<Location, Vec<BorrowIndex>>,
 
     /// NLL region inference context with which NLL queries should be resolved
-    nonlexical_regioncx: Rc<RegionInferenceContext<'tcx>>,
+    _nonlexical_regioncx: Rc<RegionInferenceContext<'tcx>>,
+}
+
+fn precompute_borrows_out_of_scope<'a, 'tcx>(
+    mir: &'a Mir<'tcx>,
+    regioncx: &Rc<RegionInferenceContext<'tcx>>,
+    borrows_out_of_scope_at_location: &mut FxHashMap<Location, Vec<BorrowIndex>>,
+    borrow_index: BorrowIndex,
+    borrow_region: RegionVid,
+    location: Location,
+    visited_locations: &mut Vec<Location>
+) {
+    // Check if we have already visited this location and skip
+    // it if we have - avoids infinite loops.
+    if visited_locations.contains(&location) { return; }
+    visited_locations.push(location.clone());
+
+    // Next, add the borrow index to the current location's vector if the region does
+    // not contain the point at that location (or create a new vector if required).
+    if !regioncx.region_contains_point(borrow_region, location) {
+        borrows_out_of_scope_at_location
+            .entry(location.clone())
+            .and_modify(|m| m.push(borrow_index))
+            .or_insert(vec![ borrow_index ]);
+    }
+
+    let bb_data = &mir[location.block];
+    // If we are past the last statement, then check the terminator
+    // to determine which location to proceed to.
+    if location.statement_index == bb_data.statements.len() {
+        if let Some(ref terminator) = bb_data.terminator {
```

Comment:
you can invoke [the `successors` method](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/mir/struct.Terminator.html#method.successors) on the terminator instead, would be easier =)